Emphasize that Bigarray.int refers to the OCaml int type, and not the C int type

edwintorok opened the pull request:
This is already documented in 'element kinds', but I think it is worth emphasizing, because it can lead to bugs like  https://github.com/yallop/ocaml-ctypes/issues/744 if the C and OCaml types for 'int' are mixed up.

On Linux x86-64 they are of different sizes (8 bytes vs 4), and this can lead to memory corruption if one is not careful,
in fact the closest C type here would be `size_t`, and there is no bigarray that would map a C `int*` in a platform-independent way.

edwintorok committed:
doc(Bigarray.int): add link to element kinds

Signed-off-by: Edwin Török <edwin@etorok.net>

edwintorok committed:
doc(Bigarray.int): clarify that it refers to OCaml ints and NOT C int

This is mentioned in the 'Element kinds' documentation, but that is not
visible anywhere in LSP documentation completions.
It is worth emphasizing this, because getting this wrong may lead to
memory corruption.

Suggested-by: Florian Angeletti <<florian.angeletti@inria.fr>>
Signed-off-by: Edwin Török <edwin@etorok.net>

Octachron commented on stdlib/bigarray.mli:
```diff
@@ -174,7 +174,9 @@ val int16_unsigned : (int, int16_unsigned_elt) kind
 
 val int : (int, int_elt) kind
-(** See {!Bigarray.char}. *)
+(** See {!Bigarray.char} and {!section:1_Elementkinds}.
```
The section label is `elementkinds`. I am not  sure if it makes sense to update only this description.

Octachron commented on stdlib/bigarray.mli:
```diff
@@ -174,7 +174,9 @@ val int16_unsigned : (int, int16_unsigned_elt) kind
 val int : (int, int_elt) kind
-(** See {!Bigarray.char}. *)
+(** See {!Bigarray.char} and {!section:1_Elementkinds}.
+    The element size matches the OCaml [int] type, and not the C [int] type.
```
I think you can simply repeat the description that this is the kind of OCaml integers array.

> Beware that this kind represents an array of OCaml [int], and not an array of 32 bits [int] a la C.

edwintorok replied:
I've copied the description from the element kinds, and simplified the reference to `C`: "they do not match the C [int] type".

The other element kinds have names that are very similar to types from C, which is probably where the confusion comes from. Emphasizing the distinction might be useful, given that it has already led to bugs in projects like `ctypes` (see original description).

edwintorok commented:
Sorry, this PR must've gotten lost in my inbox. I've updated it now, I think it also needs the `no-change-entry-needed` label, unless you want me to add a Changes entry for it?

Octachron commented on stdlib/bigarray.mli:
```diff
@@ -186,7 +186,12 @@ val int16_unsigned : (int, int16_unsigned_elt) kind
+
+ OCaml integers (signed, 31 bits on 32-bit architectures,
+   63 bits on 64-bit architectures).
+ They do not match the C [int] type.
```
Let be more explicit: since we are writing down a warning, it should looks like a warning even for people which do not have the context of this PR.

> Beware that this a bigarray containing OCaml integers (signed, 31 bits on 32-bit architectures,
   63 bits on 64-bit architectures), which does not match the [C] int  type.

Octachron commented:
I would prefer if you added a Changes entry, documentation improvements are important improvements.

edwintorok committed:
doc(Bigarray.int): add changes entry

Signed-off-by: Edwin Török <edwin.torok@cloud.com>

edwintorok commented:
I fixed the conflicts in `Changes`, I think the PR numbers are meant to be sorted in each section, so I kept this one as first in the 'Manual and documentation' section and moved 13295 last.

Octachron commented:
The ordering of Changes entries is not strictly enforced, but moving around the change entry looks fine.
Finally merged, thanks for your patience!
